pkg/couchdb/stream/all_docs.go

```diff
@@ -0,0 +1,350 @@
+	}
+}
+
+func (f *allDocsFilter) Float(x float64) {
+	f.value(x)
+}
+
+func (f *allDocsFilter) Number(n string) {
+	if f.err == nil {
+		f.err = fmt.Errorf("number %q is not supported", n)
+	}
+}
+
+func (f *allDocsFilter) String(s string) {
+	if f.skipDDoc && f.depth == 3 && f.path == "id" && strings.HasPrefix(s, "_design") {
+		// skip design docs
+		f.rowIsDDoc = true
+		f.path = ""
+	} else {
+		f.value(s)
+	}
+}
+
+func (f *allDocsFilter) Key(s string) {
+	if f.path == "" {
+		f.path = s
+	} else {
+		f.path += "." + s
+	}
+}
+
+func (f *allDocsFilter) ObjectStart() {
+	var err error
+	switch f.depth {
+	case 0: // global
+		// nothing
+	case 1: // rows array
+		err = errors.New("unexpected case")
+	case 2: // a row
+		f.rowIsDDoc = false
+		f.path = ""
+		err = f.row.Object()
+	case 3: // doc or value
+		if len(f.fields) == 0 || f.path != "value" {
+			err = f.row.Object(f.path)
+		}
+	default: // inside doc
+		err = f.objectStartInDoc()
+	}
+	if err != nil && f.err == nil {
+		f.err = err
+	}
+	f.depth++
+}
+
+func (f *allDocsFilter) objectStartInDoc() error {
+	// We are inside an object that won't be copied to the response
+	if f.rejectedAt >= 0 {
+		return nil
+	}
+
+	// Objects inside an array are always kept
+	key := f.currentKey()
+	if key == "[]" {
+		return f.row.Object()
+	}
+
+	// We need to keep cozyMetadata.uploadedBy if fields include cozyMetadata,
+	// and we keep everything if fields is empty.
+	if f.matchedAt >= 0 || len(f.fields) == 0 {
+		return f.row.Object(key)
+	}
+
+	// Exact match
+	for _, field := range f.fields {
+		if field == f.path {
+			f.matchedAt = f.depth
+			return f.row.Object(key)
+		}
+	}
+
+	// We need to keep metadata if fields include metadata.datetime
+	for _, field := range f.fields {
+		if strings.HasPrefix(field, f.path+".") {
+			return f.row.Object(key)
+		}
+	}
+
+	// We can remove this object from the response
+	f.rejectedAt = f.depth
+	return nil
+}
+
+func (f *allDocsFilter) ObjectEnd() {
+	f.depth--
+
+	switch f.depth {
+	case 0: // global
+		// nothing
+	case 1: // rows array
+		if f.err == nil {
+			f.err = errors.New("unexpected case")
+		}
+	case 2: // a row
+		if f.rowIsDDoc {
+			f.row.Reset()
+		} else {
+			f.flushRow()
+		}
+	case 3: // doc or value
+		if len(f.fields) == 0 || f.path != "value" {
+			f.row.Pop()
+		}
+		f.path = ""
+	default: // inside doc
+		f.objectEndInDoc()
+	}
+}
+
+func (f *allDocsFilter) objectEndInDoc() {
+	if key := f.currentKey(); key != "[]" {
+		f.popKey(key)
+	}
+
+	if f.rejectedAt >= 0 {
+		if f.rejectedAt == f.depth {
+			f.rejectedAt = -1
+		}
+		return
+	}
+	if f.matchedAt == f.depth {
+		f.matchedAt = -1
+	}
+
+	f.row.Pop()
+}
+
+func (f *allDocsFilter) flushRow() {
+	prefix := ""
+	if f.total != 0 {
+		prefix = ","
+	}
+	row := prefix + oj.JSON(f.row.Result()) + "\n"
+	f.row.Reset()
+	if _, err := f.w.Write([]byte(row)); err != nil && f.err != nil {
+		f.err = err
+	}
+	f.total++
+}
+
+func (f *allDocsFilter) ArrayStart() {
+	f.depth++
```

Comment:
Why incrementing the depth at the beginning of the function when it is increment only at the end of `ObjectStart`?

Reply:
The "logical" place would be to put the `f.depth++` at the end of the function. But we have several early returns, and we would need to duplicate this line for each returns. So, I have preferred to keep only one line for incrementing the depth, even if it is not the "logical" place, but we could go with the other option if you prefer (I don't really mind).